route_auth.go

```diff
@@ -1,102 +1,242 @@
-					SetEmail(r.PostFormValue("email")).
-					SetPassword(string(password)).
-					Save(r.Context()) // Create and return.
-
-		// TODO: Check availability of email address.
-		// TODO: Validate inputs.
+		// Don't create an account for the user if an account with this email already exists.
+		// For privacy reasons we can't respond with a form error because that would disclose
+		// who registered to the game and who didn't. Instead we can e-mail the user a link
+		// to reset their password.
+		// TODO: first implement a password reset. Then send a modified password reset email
+		//       when the user tries to register with an email that is already used. But
+		//       change the message a bit so that it is clear why they received a password
+		//       reset instead of an activation email and that they can ignore the email
+		//       if it wasn't them.
+		n, err := tx.User.
+			Query().
+			Where(user.EmailEQ(input.Email)).
+			Count(r.Context())
+		if n > 0 {
+			return nil
+		}
+
+		u, err := tx.User.
+			Create().
+			SetUsername(input.Username).
+			SetEmail(input.Email).
+			SetPassword(string(hashedPassword)).
+			SetVerifyToken(verifyToken).
+			Save(r.Context())
 		if err != nil {
 			return err
 		}
 
 		c, err := tx.Planet.Query().
-				Where(
-					planet.And(
-						planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
-						planet.Not(planet.HasOwner()),
-					),
-				).
-				Count(r.Context())
+			Where(
+				planet.And(
+					planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
+					planet.Not(planet.HasOwner()),
+				),
+			).
+			Count(r.Context())
 
 		if err != nil {
 			return err
 		}
 
-		ra := rand.New(rand.NewSource(42))
-
-		n := ra.Intn(c)
+		ra := rand.New(rand.NewSource(time.Now().UnixNano()))
+		rn := ra.Intn(c)
 
 		p, err := tx.Planet.Query().
-				Where(
-					planet.And(
-						planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
-						planet.Not(planet.HasOwner()),
-					),
-				).
-				Offset(n).
-				First(r.Context())
+			Where(
+				planet.And(
+					planet.PlanetTypeEQ(planet.PlanetTypeHabitable),
+					planet.Not(planet.HasOwner()),
+				),
+			).
+			Offset(rn).
+			First(r.Context())
 
 		_, err = p.Update().
 			SetOwner(u).
 			Save(r.Context())
 
+		if err := sendSignupEmail(u); err != nil {
+			return err
+		}
+
 		return err
 	})
 	if err != nil {
 		serveError(w, r, newInternalServerError(fmt.Errorf("unable to create a user account: %v", err)))
 		return
 	}
 
-	flash(r, flashSuccess, "Your account has been created. You can log in now.")
-	http.Redirect(w, r, "/", 302)
+	// TODO: Move this to a separate page so the user isn't tempted to try to
+	//       log in even through they haven't activated their account yet.
+	flash(r, flashSuccess, "We have send you an email with a confirmation link")
+	http.Redirect(w, r, "/", 303)
+}
+
+var generateVerifyToken = func() (string, error) {
+	return generateRandomString(20)
+}
+
+func sendSignupEmail(u *ent.User) error {
```

Comment:
Shouldnt this function be in the mail.go file?

Reply:
Could be in both. I will move it.